src/dmd/root/file.d

```diff
@@ -22,6 +22,379 @@ import dmd.root.filename;
 import dmd.root.rmem;
 import dmd.root.string;
 
+/**
+Encapsulated management of a memory-mapped file.
+
+The `Datum` type encodes the mapping mode: Use `ubyte` for read/write mapping
+and `const ubyte` for read-only mapping.
+*/
+struct FileMapping(Datum)
+{
+    static assert(Datum.sizeof == 1);
+
+    version(Posix) enum invalidHandle = -1;
+    else version(Windows) enum invalidHandle = INVALID_HANDLE_VALUE;
+
+    /// Handle of underlying file
+    private auto handle = invalidHandle;
+    /// File mapping object needed on Windows
+    version(Windows) private HANDLE fileMappingObject = invalidHandle;
+    /// Memory-mapped array
+    private Datum[] data;
+    /// Name of underlying file, zero-terminated
+    private const(char)* name;
+
+    import dmd.backend.errors : error;
```

Comment:
This is adding a dependency to the backend. dmd.root is also used by the frontend, so this might make downstream backends depend on the dmd backend.

As far as I could tell, you can use printf or assert to error out.

Reply:
fixed, thanks